Polarizable Embedding model

maxscheurer opened the pull request:
## Description
This PR's purpose is to add the Polarizable Embedding (PE) model to Psi4.

In the [PE](https://pubs.acs.org/doi/10.1021/ct1003803) model, the environment is modeled by a multi-center multipole expansion with dipole-dipole polarizabilities placed at the expansion sites (similar to EFP). A tutorial on PE in general can be found [here](https://onlinelibrary.wiley.com/doi/full/10.1002/qua.25717).

I added the PE model through my [CPPE](https://github.com/maxscheurer/cppe) library on the Python level.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Ground-state (HF and DFT) SCF calculations with PE
~~- [ ] Perturbative state-specific (ptSS) corrections for EOM-CC calculations~~ can be done in user code on the Python layer
- [x] rebase once #1657 is merged
- [x] include `cppe` on the Python layer
- [x] remove C++ code
- [x] cleanup CMake
- [x] pin a released CPPE version
- [x] bump pybind to v2.3.0
- [x] simple documentation

## Questions
- [x] The tests need a _potential_ input file (`*.pot`), currently residing in the `tests/cppe/scf-hf` folder. The path specification I have to put in the input file is currently a bit weird because the tests are not run from inside the test directory. Is there any nice solution to this issue?

## Checklist
- [x] Tests added for any new features (SCF)
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

loriab commented on tests/cppe/scf-hf/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+include(TestingMacros)
+
+add_regression_test(cppe-scf-hf "psi;cppe;addon;scf")
```
Use commands like https://github.com/psi4/psi4/blob/master/tests/cubeprop/CMakeLists.txt here to get your `*.pot` file available at test-time.

maxscheurer replied:
Addressed in 1930d51b585da776121e44a700b166c00765a99d.
Thanks!

loriab commented on CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.3 FATAL_ERROR)  # 3.3: EP_add(DEPENDS interface
                                                  # 3.1: CMAKE_CXX_STANDARD, BUILD_ALWAYS
 project(psi4
-        LANGUAGES C CXX)
+	LANGUAGES C CXX Fortran)
```
Don't add `Fortran` here. Instead add your pkg to https://github.com/psi4/psi4/pull/1401/files#diff-af3b638bc2a3e6c650974192a53c7291R106

maxscheurer replied:
Oops, I don't remember why I even put this Fortran flag there...
Resolved in a030a01c6.

loriab commented on CMakeLists.txt:
```diff
@@ -85,6 +85,7 @@ option(ENABLE_ambit "Enables the ambit tensor library" OFF)
 option(ENABLE_CheMPS2 "Enables CheMPS2 for DMRG (requires HDF5)" OFF)
 option(ENABLE_dkh "Enables DKH integrals (requires Fortran)" OFF)
 option(ENABLE_libefp "Enables LIBEFP and PylibEFP for fragments" OFF)
+option(ENABLE_cppe "Enables CPPE for Polarizable Embedding" OFF)
```
alphabetize, please, ignoring "lib"

maxscheurer replied:
Resolved in a030a01c6.

loriab commented on external/upstream/cppe/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+            # UPDATE_COMMAND ""
+            CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${STAGED_INSTALL_PREFIX}
+                       -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+                       -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
```
Fortran compiler needed?

loriab commented on psi4/driver/procrouting/proc.py:
```diff
@@ -1371,6 +1371,42 @@ def scf_helper(name, post_scf=True, **kwargs):
+                pol_embed_border_options.border_type = core.PeBorderOptions.BorderType.redist
+            pol_embed_border_options.rmin = core.get_local_option('PE', 'BORDER_RMIN')
+            if core.get_local_option('PE', 'BORDER_RMIN_UNIT').upper() == "AA":
+                pol_embed_border_options.rmin *= 1.8897261246
```
`constants.bohr2angstroms`

loriab commented on psi4/src/psi4/libpsipe/pe_calc_handler_i.hh:
```diff
@@ -0,0 +1,31 @@
+#ifndef LIBPSIPE_PE_CALC_HANDLER_H
```
Please switch to `.h` extension. And avoid "psi" in dirname -- we've got so many already :-)

maxscheurer replied:
Resolved in 893f823f7 and 01ae0e5ac.
The header file you mentioned was not being used and I just forgot to remove it...
I renamed the `libpsipe` directory to `libpe`.

loriab commented on psi4/src/read_options.cc:
```diff
@@ -248,6 +248,40 @@ int read_options(const std::string &name, Options & options, bool suppress_print
+
+    /*- Name of the potential file -*/
+    options.add_str_i("POTFILE", "potfile.pot");
+    /* - Use DIIS acceleration to obtain induced moments -*/
```
Make sure the comment delimiters are exactly `/*-` and `-*/`. The scraper is picky

loriab commented on psi4/src/read_options.cc:
```diff
@@ -248,6 +248,40 @@ int read_options(const std::string &name, Options & options, bool suppress_print
+    /* - Activate border options for sites in proximity to the QM/MM border -*/
+    options.add_bool("BORDER", false);
+    /* - border type, either remove or redistribute moments/polarizabilities -*/
+    options.add_str_i("BORDER_TYPE", "REMOVE", "REMOVE REDIST");
```
Avoid the case-preserving `str_i` except where it's really needed, like filenames.

loriab commented on psi4/src/psi4/libpsipe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+
+#include <iomanip>
+
+#include <armadillo>
```
Where is armadillo coming from in the build process? How onerous would it be to switch out for LAPACK calls or `psi4.core.Matrix`?

maxscheurer replied:
When the CPPE library is built, cmake searches for armadillo on the system, and builds it from source if it's not found ([here is the code in my repo](https://github.com/maxscheurer/cppe/blob/master/external/upstream/armadillo/CMakeLists.txt)). The stand-alone CPPE build works fine on Travis CI at least.

I think at the moment it is very hard to replace armadillo since this would require re-writing a major part of the library, but I understand that having yet another linear algebra library
is annoying...

Just some additional information: I originally developed the CPPE library for PE in Q-Chem where armadillo commonly used, so it was kind of convenient to use it for CPPE, too.

dgasmith commented on psi4/src/export_cppe.cc:
```diff
@@ -0,0 +1,73 @@
+using namespace psi;
+#ifdef USING_cppe
+
+void export_cppe(py::module& m) {
```
@loriab What do you think about name spacing such extensions under `core.cppe.*`?

loriab replied:
I haven't thought about such matter, but it feels sensible. So cautious yes. Though I don't think we want to go explicit subpackage, `import psi4.core.cppe` with separate compiled lib, as @robertodr was having some problems with structuring that.

robertodr replied:
@loriab I managed to solve the problem, so it can be done without too much effort.

dgasmith commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+// TODO: implement output stream option in CPPE for printing
+
+namespace psi {
+    
```
Can you add a `cppe` namespace or similar here as well?

dgasmith commented:
Can you expand a bit on "not really header-only"? Checking the source code, I only see `hpp` files.

maxscheurer commented:
> Can you expand a bit on "not really header-only"? Checking the source code, I only see `hpp` files.

Sure. If the armadillo runtime library is to be used, `src/wrapper.cpp` is compiled and
one can simply link against `armadillo`. However, this wrapper can also be switched off. In this case, one needs to link directly against BLAS and LAPACK. The [documentation](http://arma.sourceforge.net/faq.html) explains this in a little more detail in the _Linking_ section.

I have never tried to use armadillo without the wrapper to be honest... I'll give it a try.

robertodr commented on external/upstream/cppe/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+                       -DCMAKE_INSTALL_INCLUDEDIR=${CMAKE_INSTALL_INCLUDEDIR}
+                       -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS}
+                       -DENABLE_OPENMP=${ENABLE_OPENMP}
+                       -DENABLE_PYTHON_INTERFFACE=OFF
```
I think you have a typo in this option.

maxscheurer replied:
Corrected in e8ab7d0b0e153ff326918521bd26cc13d738b481.

robertodr commented on psi4/driver/procrouting/proc.py:
```diff
@@ -1371,6 +1371,42 @@ def scf_helper(name, post_scf=True, **kwargs):
         core.print_out("""  PCM does not make use of molecular symmetry: """
                        """further calculations in C1 point group.\n""")
         use_c1 = True
+    
```
Since a 3-layer QM/MM/PCM is currently not possible, we need to check that both options are not set at the same time and abort if that's the case.

maxscheurer replied:
Resolved in c3cd34779.

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+
+SharedMatrix PeIntegralHelper::compute_multipole_potential_integrals(Vector3 site, int order, std::vector<double>& moments) {
+    auto integrals = std::make_shared<IntegralFactory>(basisset_, basisset_, basisset_, basisset_);
+    std::shared_ptr<OneBodyAOInt> multipole_integrals = static_cast<std::shared_ptr<OneBodyAOInt>>(integrals->ao_efp_multipole_potential());
```
I think `auto multipole_integrals = std::make_shared<OneBodyAOInt>(integrals->ao_efp_multipole_potential());` should suffice here.

maxscheurer replied:
Done in b9742ad4d.

robertodr commented on psi4/src/psi4/libmints/oeprop.cc:
```diff
@@ -826,6 +827,39 @@ void OEProp::compute() {
+    if (tasks_.count("PE_CORR")) compute_pe();
+}
+
+PePropCalc::PeOutputType PePropCalc::compute_pe_prop(bool transition, bool print_output, bool verbose) {
```
I don't think this belongs here as a one-electron property. It also seems the only thing you need to compute the perturbative correction is the AO density matrix, hence this can possibly become a free function in the Python layer of the code?

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+    fields.push_back(std::make_shared<Matrix>("AO Field Z", basisset_->nbf(), basisset_->nbf()));
+    
+    auto integrals = std::make_shared<IntegralFactory>(basisset_, basisset_, basisset_, basisset_);
+    std::shared_ptr<OneBodyAOInt> field_integrals_ = static_cast<std::shared_ptr<OneBodyAOInt>>(integrals->electric_field());
```
I think `auto field_integrals_ = std::make_shared<OneBodyAOInt>(integrals->electric_field());` should suffice here too.

maxscheurer replied:
Resolved in b9742ad4d.

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+            }
+            V_es_->add(int_helper_.compute_multipole_potential_integrals(site, 2, moments));
+        }
+        cppe_state_.set_es_operator(arma::mat(V_es_->get_pointer(), V_es_->nrow(), V_es_->ncol(), true));
```
Would it be possible to add an overload for this function in CPPE that accepts the raw pointer and the sizes and reconstructs the Armadillo matrix internally? It could also take care of saving the matrix to text file. 

I guess this could be a good strategy to avoid Armadillo spilling into Psi4 (or other QC programs, for that matter)

robertodr replied:
Also, why do you need to keep track of the electrostatics operator inside the library?

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+        cppe_state_.es_operator_copy().save("V_es_psi4.txt",  arma::raw_ascii);
+    }
+    
+    SharedMatrix D = Dm;
```
I am not positive you need to keep a copy of the SCF density matrix at hand here. For example, in CC you can pass in the correlation-only density matrix and set the `CalcType` to `electronic_only` to achieve the same effect (I think there's an outdated PR of mine showing that for PCM). And if that's not feasible, the wavefunction itself carries the current density matrix, so you can do the subtraction _outside_ of here and simplify the state of the `PeState` object a bit.

maxscheurer replied:
I agree. Removed it in 22bb0665c.

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+    if (subtract_scf_density && type == PeState::CalcType::electronic_only) {
+        D->subtract(D_scf_);
+    }
+    cppe_state_.update_energies(arma::mat(D->get_pointer(), D->nrow(), D->ncol(), true));
```
Having not read the CPPE code, I think the naming of this function is a bit confusing. If it updates energies, why does it take the density matrix as argument? Is it because you trace the operator matrices with the density matrix internally to get the embedding energy? Couldn't that be done Psi4-side?

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+    // TODO: only for debugging
+    // cppe_state_.print_summary();
+    
+    if (type == PeState::CalcType::total) iteration++;
```
The iteration number could be passed from outside when calling this function, rather than being kept track of here. Why does it only increment when `CalcType` is `total`? In some CC models with polarization you also need to iterate until self-consistency, but `CalcType` would be `electronic_only`.

maxscheurer replied:
Resolved in be3006403.

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+    field_integrals_->set_origin(site);
+    field_integrals_->compute(fields);
+    
+    fields[0]->scale(-1.0*moment[0]);
```
A bit more compact:
```
for (int i = 0; i < 3; ++i) {
   fields[i]->scale(-1.0*moment[i]);
}
```

maxscheurer replied:
Done in b9742ad4d.

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+    field_integrals_->compute(fields);
+    
+    Vector efields("efields", 3);
+    efields.set(0, D->vector_dot(fields[0]));
```
A bit more compact:
```
for (int i = 0; i < 3; ++i) {
  efields[i] = D->vector_dot(fields[i]);
}
```

maxscheurer replied:
Done in b9742ad4d.

robertodr commented on psi4/src/psi4/libpe/psipe.h:
```diff
@@ -0,0 +1,110 @@
+    // void calculate_excited_state_energy_correction(double* p_exc, double* energy, bool is_tdm);
+
+  private:
+      std::shared_ptr<BasisSet> basisset_;
```
I guess once the `PeIntegralHelper` is initialized, it will hold the `basisset_` and there's probably no need for a redundant instance in this class?

robertodr commented on psi4/src/psi4/libpe/psipe.cc:
```diff
@@ -0,0 +1,259 @@
+
+std::pair<double, SharedMatrix> PeState::compute_pe_contribution(const SharedMatrix &Dm, CalcType type, bool subtract_scf_density) {
+    // build the electrostatics operator only once!
+    if (!iteration && (type == PeState::CalcType::total)) {
```
If this is built once and stored, it's probalby better to do it in the CTOR.

robertodr commented on psi4/src/psi4/libpe/psipe.h:
```diff
@@ -0,0 +1,110 @@
+  public:
+    enum CalcType { total, electronic_only };
+    PeState() = default;
+    PeState(libcppe::PeOptions options, std::shared_ptr<BasisSet> basisset);
```
This CTOR can be marked `explicit`.

maxscheurer replied:
Done in b9742ad4d.

robertodr commented on psi4/src/psi4/libpe/psipe.h:
```diff
@@ -0,0 +1,110 @@
+      PeIntegralHelper int_helper_;
+      
+      SharedMatrix V_es_;
+      SharedMatrix D_scf_;
```
As per comments above, `D_scf_` and `iteration` are most likely redundant.

susilehtola commented:
Armadillo can be used without the wrapper without any problems. Personally I have problems understanding why the wrapper was implemented in the first place, it usually just is in the way. All that the wrapper does is link to the BLAS and LAPACK library.

maxscheurer commented:
@robertodr Thanks for the detailed code review. I have to admit that I obviously did not clean up the code properly, sorry for that. Also, my CPPE library is still in an early stage and under heavy development, so thanks for pointing out some design/naming flaws. I'll try to address your points as soon as I find the time.

dgasmith commented:
@robertodr @loriab Can we get a fresh look here when you have a chance?

jturney commented on psi4/src/psi4/libmints/multipolepotential.cc:
```diff
@@ -0,0 +1,221 @@
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/integral.h"
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
```
I can't remember. Feel free to change to the `std::` versions.

maxscheurer commented:
Since `cppe` has its own Python interface now (which it did not when I started this PR), I guess there is no reason to interface with psi4 on the C++ layer?

I guess it would be much simpler wrt maintenance of the code to refactor everything and use the `cppe` Python interface directly. What do you think, @loriab and @dgasmith ?

dgasmith commented:
I think that is likely the way to go, I believe you will still need to merge some of the new integral code. It may be worth splitting that off into a another PR to get a digestible component there.

We usually can merge small chunks quicker and expanding functionality through python is usually quite preferred if that level of integration is sufficient.

maxscheurer commented:
Okay, then I'll go for it.

> It may be worth splitting that off into a another PR to get a digestible component there.
Alright.

maxscheurer commented:
I added some experimental changes to `electricfield.{cc,h}` in order to compute
field integrals (for the induction operator) and electric field expectation values (needed to solve for induced moments) for multiple sites/coordinates at once.

Even though I don't like the code I implemented, it gives massive performance improvements.
In my tests (Python code is uploaded [here](https://gist.github.com/maxscheurer/4736a100ff5de12190ef27f7bc2eaade)), I computed integrals at 2000 random points for Water/cc-pVQZ, for which I get a 25x speed-up compared to the existing implementation. EFP would also benefit from these performance improvements.

Since the code is rather ugly at the moment, I'm thankful for any suggestions on how to proceed from here.

dgasmith commented:
The integral speedups would look awesome, but would be good in another PR. The more digestible the better! If you could split this off we could continue this conversation in the new PR.

maxscheurer commented:
> The integral speedups would look awesome, but would be good in another PR. The more digestible the better! If you could split this off we could continue this conversation in the new PR.

Sounds like a good plan.

dgasmith commented:
Awesome, if you pull that out. Is this PR ready for review and merge?

maxscheurer commented:
No, I have to refactor the code such that the PE stuff is Python-side only (see ToDo list above).
In the meantime, I'll hopefully a) have a first "release" of `cppe` ready that can be pinned in the `ExternalProject` and b) have the PR with performance improvements for electric field integrals done.

Is there a deadline for the Psi4 1.4 release?

dgasmith commented:
I think we have a general target of this fall for Psi4 1.4

maxscheurer commented:
> I think we have a general target of this fall for Psi4 1.4

Alright, that sounds feasible :blush:

maxscheurer commented:
**Update**

I've completely replaced the C++-based implementation of PE with a light-weight Python-side version.
Further, I had to bump pybind11 to `v.2.2.4` for the CPPE library to work correctly, and I pinned a release version of CPPE in the `ExternalProject`.

So I guess this PR will be ready for merge soon. I'm going to add some simple documentation now.

robertodr commented on psi4/driver/procrouting/proc.py:
```diff
@@ -54,6 +54,15 @@
 from . import mcscf
 from . import response
 
+_have_cppe = False
```
Doesn't the same portion of code in `solvent/__init__.py` suffice?

maxscheurer replied:
Yes. Done.

maxscheurer commented:
Basic documentation is now added.

I think the only thing missing is to push a recent version of CPPE (0.0.8) to conda, @loriab .
Other than that, this is ready for merge, in my opinion.

loriab commented on external/upstream/cppe/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+                       -DCMAKE_INSTALL_DATADIR=${CMAKE_INSTALL_DATADIR}
+                       -DCMAKE_INSTALL_INCLUDEDIR=${CMAKE_INSTALL_INCLUDEDIR}
+                       -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS}
+                       -DENABLE_OPENMP=${ENABLE_OPENMP}
```
How much of a performance hit would turning off OpenMP (in cppe, not psi4) be? Not permanently but as a temporary, let's-get-this-integrated measure.

I suggest this because https://github.com/maxscheurer/cppe/blob/master/cppe/CMakeLists.txt#L12-L18 is fine for standalone, but it's going to choose a threading library based only on compiler, not taking LAPACK threading or `OpenMP_LIBRARY_DIRS` into account. Psi uses the crazy [`MathOpenMP`](https://github.com/psi4/psi4/blob/master/external/common/lapack/FindMathOpenMP.cmake) to get this right, a scheme I've no affection for but guardedly trust. We can find a way to hand the correct openmp flags to cppe (which is how we do lapack for other addons), but I don't want that detail to hold up this PR.

maxscheurer replied:
Won't be an issue I guess. One can always provide a self-built `cppe` via `PYTHONPATH` that is built with OpenMP for the time being.
Will turn it off.

maxscheurer replied:
Done.

loriab commented on external/upstream/pybind11/CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@ else()
     message(STATUS "Suitable pybind11 could not be located, ${Magenta}Building pybind11${ColourReset} instead.")
     ExternalProject_Add(pybind11_external
-        URL https://github.com/pybind/pybind11/archive/v2.2.3.tar.gz
+        URL https://github.com/pybind/pybind11/archive/v2.2.4.tar.gz
```
for consistency, bump it [here](https://github.com/psi4/psi4/pull/1401/files#diff-e8cf7e78ee17d1826f8f9cef1a04c7bfR60), too, please.

maxscheurer replied:
Done.

loriab commented on psi4/__init__.py:
```diff
@@ -100,5 +100,7 @@
 if "@ENABLE_PCMSolver@".upper() in ["1", "ON", "YES", "TRUE", "Y"]:
     sys.path.insert(1, "@PCMSolver_PYMOD@")
+if "@ENABLE_cppe@".upper() in ["1", "ON", "YES", "TRUE", "Y"]:
+    sys.path.insert(1, "@cppe_PYMOD@")
```
this `_PYMOD` variable is provided by you, akin to https://github.com/PCMSolver/pcmsolver/blob/master/cmake/PCMSolverConfig.cmake.in#L143. I don't think that's happening?

maxscheurer replied:
Will be done in cppe 0.0.9.

maxscheurer replied:
Done.

loriab commented on psi4/driver/procrouting/solvent/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+_have_pe = False
+try:
+    from . import pol_embed
```
won't the `pol_embed` file always be present?

maxscheurer replied:
Yes, but inside `pol_embed`, `cppe` is imported. So if the top-level `pol_embed` import fails, PE won't be available.

loriab commented on psi4/driver/procrouting/solvent/pol_embed.py:
```diff
@@ -0,0 +1,120 @@
+    mol = cppe.Molecule()
+    geom = psi4mol.geometry().np
+    for i, c in enumerate(geom):
+        a = cppe.Atom((int(psi4mol.Z(i))), *c)
```
be good to try this on something with a ghost atom (`Z(i)` will be 0.0) to make sure nothing breaks and result is sane.

maxscheurer replied:
Result looks good with a ghost atom.

loriab commented on psi4/src/read_options.cc:
```diff
@@ -260,6 +260,42 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        /*- Use DIIS acceleration to obtain induced moments -*/
+        options.add_bool("DIIS", true);
+        /*- Threshold for induced moments convergence -*/
+        options.add_int("CONVERGENCE_INDUCED", 8);
```
ok to reverse this to `INDUCED_CONVERGENCE` to be more predictable wrt `D_CONVERGENCE`, etc.?

maxscheurer replied:
Done.

loriab commented on tests/CMakeLists.txt:
```diff
@@ -44,7 +44,7 @@ foreach(test_name adc1 adc2 casscf-fzc-sp casscf-semi casscf-sa-sp ao-casscf-sp
                   cc9 cc9a cdomp2-1 cdomp2-2 cepa0-grad1 cepa0-grad2 cepa1
                   cepa2 cepa3 cepa4 cepa-module ci-multi cisd-h2o+-0 cisd-h2o+-1
-                  cisd-h2o+-2 cisd-h2o-clpse cisd-opt-fd cisd-sp cisd-sp-2
+                  cisd-h2o+-2 cisd-h2o-clpse cisd-opt-fd cisd-sp cisd-sp-2 cppe
```
ah, this doesn't go here but instead the `cppe/` dir gets added like https://github.com/psi4/psi4/pull/1401/files#diff-3dea97423f83302e647a3bab62bf1157L157

maxscheurer replied:
Don't know what line the link refers to :slightly_frowning_face: 
Could you specify a line number?

loriab replied:
https://github.com/psi4/psi4/blob/master/tests/CMakeLists.txt#L155-L161

maxscheurer replied:
Done.

loriab commented on psi4/driver/procrouting/proc.py:
```diff
@@ -53,6 +53,7 @@
 from . import dft
 from . import mcscf
 from . import response
+from . import solvent
```
note for a second review pass: need to find a spot (scf_helper?) where we look if `PE` is on and if so, set c1/noreorient so user doesn't need to handle this.

maxscheurer replied:
I'll try to implement a clean way of doing this... the problem is that PE needs information about the molecule and basis set to construct the `MintsHelper` downstream... So one would have to do it [here](https://github.com/psi4/psi4/blob/083fb200b45aa3e0390b934b23627d83d9abf3bc/psi4/driver/procrouting/proc.py#L1500), where the molecule/basisset are already transformed to C1.

maxscheurer replied:
Okay, I give up on this for today...

maxscheurer replied:
I guess the only way to do this is through the user input. The molecule must not be moved or rotated... and there is no way of knowing if PE is enabled at the moment the molecule is constructed. Any suggestions?

loriab replied:
Unless PE uses are peculiar, don’t need to know c1ness at mol creation time, only need to set before computation time. Mol class stores original and applies transformations.

maxscheurer replied:
I'm giving up on this... probably you can help, @loriab ☹️ 
I played around with code from [here](https://github.com/psi4/psi4/blob/a8b9036e49d01f385fa3ccfb4b3eb48c13785899/psi4/driver/procrouting/proc.py#L1461), but could not make it work after all...

maxscheurer commented on CMakeLists.txt:
```diff
@@ -187,6 +188,7 @@ ExternalProject_Add(psi4-core
            chemps2_external
            dkh_external
            libefp_external
+           cppe_external
```
Done.

dgasmith commented on doc/sphinxman/source/cppe.rst:
```diff
@@ -0,0 +1,213 @@
+* .. image:: https://anaconda.org/psi4/cppe/badges/version.svg
+     :target: https://anaconda.org/psi4/cppe
+
+* CPPE is available as a conda package for Linux and macOS (and Windows, through the Ubuntu shell).
```
Any chance of this compiling on native Windows? We can likely help with that and we might have some appveyor/Azure tips for you for the CI.

maxscheurer replied:
I guess it should compile on native Windows since `cppe` ships with Q-Chem, and Q-Chem is available for Windows... though I have no idea how to setup the CI for Windows. Help would be much appreciated, @dgasmith, probably in a PR for `cppe`?

dgasmith replied:
A good place to start might be [here](https://github.com/MolSSI/cookiecutter-cms/blob/master/%7B%7Bcookiecutter.repo_name%7D%7D/appveyor.yml) which is the MolSSI cookie cutter example. This does not include compiled code, but will get you setup with Appveyor. Another example is [gau2grid](https://github.com/dgasmith/gau2grid/blob/master/appveyor.yml) which has some compiled code. The two together should give you enough information. If you can get a PR together I can likely comment on it on CPPE git.

dgasmith commented:
Ready to go I think, lets get the reviews in.

loriab commented on psi4/driver/procrouting/solvent/pol_embed.py:
```diff
@@ -0,0 +1,153 @@
+
+    pol_embed_options.do_diis = core.get_local_option('PE', 'DIIS')
+    pol_embed_options.maxiter = core.get_local_option('PE', 'MAXITER')
+    pol_embed_options.pe_border = core.get_local_option('PE', 'BORDER')
```
I think you want `get_option` (same args) in these lines and below rather than `get_local_option`. Details why upon request, but you're better off not knowing. Any tests that previously worked with `set cppe {...}` should now work the same with `set {...}` (provided the option doesn't overlap and contradict a setting in a different used module). It's just a little more flexible if users aren't required to specify the local scope.

If you have experience that contradicts this, let me know.

maxscheurer replied:
I'd prefer to have the separate input section. In this manner, the input is similar to the other solvent models, e.g., PCM.

loriab replied:
A `set pe { ... }` block will still work as expected if switch to `get_option`. Right now, `set potfile pna_6w.pot` is going to do nothing, and I worry that will be confusing. But if you really want to try it, ok.

maxscheurer replied:
Done.

loriab commented on psi4/driver/procrouting/solvent/pol_embed.py:
```diff
@@ -0,0 +1,153 @@
+        V_pe = np.zeros((n_bas, n_bas))
+        if self._enable_induction:
+            # obtain expectation values of elec. field at polarizable sites
+            elec_fields = self.mints.electric_field_value(
```
If you're dedicated to this formatting, I suspect you'll want a `# yapf disable` http://psicode.org/psi4manual/master/prog_style_py.html?highlight=yapf to preserve it. Otherwise, you may want to run `yapf` over the code now to better preserve attribution.

maxscheurer replied:
I ran `yapf` in the latest commit.

maxscheurer commented on psi4/src/read_options.cc:
```diff
@@ -260,6 +260,42 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        /*- Use DIIS acceleration to obtain induced moments -*/
+        options.add_bool("DIIS", true);
+        /*- Threshold for induced moments convergence -*/
+        options.add_int("INDUCED_CONVERGENCE", 8);
```
Will be changed upstream in `cppe 0.1.1`, which I will push later today and pin in `psi4`.

maxscheurer replied:
Done.

maxscheurer replied:
Done.

maxscheurer commented on psi4/driver/procrouting/solvent/pol_embed.py:
```diff
@@ -0,0 +1,153 @@
+
+def get_pe_options():
+    if core.get_option('SCF', 'PCM'):
+        raise ValidationError("""Error: 3-layer QM/MM/PCM not implemented.\n""")
```
Done.

dgasmith commented:
@loriab can you give this a quick glance?

maxscheurer commented:
@dgasmith
I need to make some minor changes (see @loriab’s comments) before merge.

dgasmith commented:
Ah ok, I thought the latest changes had included those. Just eager to get this in :)

maxscheurer commented:
Me too... I‘ll finish up ASAP (when I get back home tomorrow evening) :smile:

maxscheurer commented:
Ready to go IMHO 😄

maxscheurer commented:
Now that all the tests pass (travis working again), I think this can be merged. ping @loriab

dgasmith commented:
@maxscheurer Awesome! Thank you for working to get this PR in. It was a new type of plugin for us, so thank you for bearing with the comments and changes.

maxscheurer commented:
I've addressed @loriab's final comments. Once the CI is finished, this can be merged.

dgasmith commented:
In, we go. Thanks again @maxscheurer!
